File: examples/kraken/src/kraken/core/envelope.py
"""Kraken Spot REST's response envelope, `{error, result}`, and its mapping onto
`truewire_core` exceptions, per the venue's own categorized error reference.
"""

from typing_extensions import Any, NotRequired
import httpx

from truewire_core.exceptions import ApiError, AuthError, BadRequest, RateLimited
from truewire_core.validation import TypedDict, validator


class Envelope(TypedDict):
  """Every Spot REST response, public or private.

  A non-empty `error` array means failure, even though the HTTP status stays 200.
  """

  error: list[str]
  result: NotRequired[Any]


validate_envelope = validator(Envelope)

CATEGORY_EXCEPTIONS: dict[str, type[ApiError]] = {
  'EAPI': AuthError,
  'EAuth': AuthError,
  'EAccount': AuthError,
  'EGeneral': BadRequest,
  'ETrade': BadRequest,
  'EFunding': BadRequest,
  'EOrder': ApiError,
  'EService': ApiError,
  'EBM': ApiError,
}
"""Default exception per error category prefix (`<Category>:<Description>`), from
Kraken's own error categorization. Overridden by the substring lists below for messages
whose meaning cuts across categories (e.g. `EOrder:Rate limit exceeded`).
"""

RATE_LIMIT_SUBSTRINGS = (
  'Rate limit exceeded',
  'Too many requests',
  'Orders limit exceeded',
  'Domain rate limit exceeded',
  'Scheduled orders limit exceeded',
)
AUTH_SUBSTRINGS = (
  'Invalid key',
  'Invalid signature',
  'Invalid nonce',
  'Permission denied',
  'Temporary lockout',
)
BAD_REQUEST_SUBSTRINGS = (
  'Invalid price',
  'Tick size check failed',
  'Order minimum not met',
  'Cost minimum not met',
)


def raise_error(errors: list[str]):
  """Raise the `truewire_core` exception matching Kraken's `error` array.

  Args:
    errors: The envelope's non-empty `error` array; only the first entry is inspected,
      matching Kraken's own convention of listing the primary failure first.

  Raises:
    AuthError: Key, signature, nonce, permission or lockout rejection.
    RateLimited: Request or order throttled.
    BadRequest: Malformed or out-of-range request parameters.
    ApiError: Any other category (order/service/business-logic failures).

  References:
    - [Error messages](https://docs.kraken.com/exchange/guides/general/errors)
  """
  message = errors[0]
  category = message.partition(':')[0]
  if any(s in message for s in RATE_LIMIT_SUBSTRINGS):
    raise RateLimited(errors)
  if any(s in message for s in AUTH_SUBSTRINGS):
    raise AuthError(errors)
  if any(s in message for s in BAD_REQUEST_SUBSTRINGS):
    raise BadRequest(errors)
  raise CATEGORY_EXCEPTIONS.get(category, ApiError)(errors)


def raise_http_status(response: httpx.Response):
  """Raise on a non-2xx HTTP status.

  Rare for Kraken -- it answers 200 even for most logical errors -- but edge/proxy
  failures (e.g. Cloudflare) still surface this way.

  Args:
    response: The unsuccessful HTTP response.

  Raises:
    AuthError: Status `401` or `403`.
    RateLimited: Status `429`.
    BadRequest: Any other `4xx` status.
    ApiError: Any other unsuccessful status.
  """
  try:
    payload: Any = response.json()
  except ValueError:
    payload = response.text
  status = response.status_code
  if status in (401, 403):
    raise AuthError(status, payload)
  if status == 429:
    raise RateLimited(status, payload)
  if 400 <= status < 500:
    raise BadRequest(status, payload)
  raise ApiError(status, payload)


def unwrap(response: httpx.Response) -> Any:
  """Return the envelope's `result`, raising on either failure discriminator.

  Args:
    response: The HTTP response to unwrap.

  Raises:
    ValidationError: The body was not JSON, or not a recognizable envelope.
    ApiError: The `error` array was non-empty, or the HTTP status was unsuccessful.
  """
  if not response.is_success:
    raise_http_status(response)
  envelope = validate_envelope(response.text)
  if envelope['error']:
    raise_error(envelope['error'])
  return envelope.get('result')

File: examples/kraken/src/kraken/core/envelope.ts
/**
 * Kraken Spot REST's `{error, result}` envelope, and its errors mapped onto the
 * `@truewire/core` hierarchy by Kraken's own error categories.
 *
 * @see https://docs.kraken.com/api/docs/guides/global-errors
 */
import { ApiError, AuthError, BadRequest, RateLimited, ValidationError } from '@truewire/core'

/** The `ApiError` subclass per `<Category>:<Description>` prefix, unless a substring below decides first. */
const CATEGORY_ERRORS: Record<string, typeof ApiError> = {
  EAPI: AuthError,
  EAuth: AuthError,
  EAccount: AuthError,
  EGeneral: BadRequest,
  ETrade: BadRequest,
  EFunding: BadRequest,
}

const RATE_LIMIT_SUBSTRINGS = [
  'Rate limit exceeded', 'Too many requests', 'Orders limit exceeded', 'Domain rate limit exceeded',
  'Scheduled orders limit exceeded',
]
const AUTH_SUBSTRINGS = ['Invalid key', 'Invalid signature', 'Invalid nonce', 'Permission denied', 'Temporary lockout']
const BAD_REQUEST_SUBSTRINGS = ['Invalid price', 'Tick size check failed', 'Order minimum not met', 'Cost minimum not met']

/** Throw the error a non-empty `error` array calls for; the first entry decides, as Kraken lists the primary failure first. */
export function raiseError(errors: string[]): never {
  const message = errors[0] ?? 'unknown error'
  const category = message.split(':', 1)[0] ?? ''
  const options = { body: errors }
  if (RATE_LIMIT_SUBSTRINGS.some(s => message.includes(s))) throw new RateLimited(message, options)
  if (AUTH_SUBSTRINGS.some(s => message.includes(s))) throw new AuthError(message, options)
  if (BAD_REQUEST_SUBSTRINGS.some(s => message.includes(s))) throw new BadRequest(message, options)
  throw new (CATEGORY_ERRORS[category] ?? ApiError)(message, options)
}

/** A non-2xx status, rare for Kraken (it answers 200 to most logical errors) but what an edge failure looks like. */
export function raiseHttpStatus(status: number, text: string): never {
  let body: unknown = text
  try { body = JSON.parse(text) } catch { /* not JSON: keep the text */ }
  const message = `HTTP ${status}: ${text.slice(0, 200)}`
  const options = { status, body }
  if (status === 401 || status === 403) throw new AuthError(message, options)
  if (status === 429) throw new RateLimited(message, options)
  if (status >= 400 && status < 500) throw new BadRequest(message, options)
  throw new ApiError(message, options)
}

/** The envelope's `result`, or the error its `error` array or status calls for. */
export function unwrap(status: number, text: string): unknown {
  if (status >= 400) raiseHttpStatus(status, text)
  let envelope: unknown
  try { envelope = JSON.parse(text) } catch (e) {
    throw new ValidationError(`invalid JSON: ${(e as Error).message}`, { cause: e })
  }
  if (typeof envelope !== 'object' || envelope === null || !Array.isArray((envelope as { error?: unknown }).error)) {
    throw new ValidationError('not a Kraken envelope: no `error` array', { issues: [{ path: '/error', message: 'expected an array' }] })
  }
  const { error, result } = envelope as { error: string[]; result?: unknown }
  if (error.length > 0) raiseError(error)
  return result
}
